### DISCUSSION TITLE
Option for disabling Copilot in specific files

### DISCUSSION THREAD

**Discussion by kam193** [community/community#11254] [Copilot Conversations]
Hey,
I'm seeking an option to disable the Copilot in specific files. I want to avoid disabling it for a whole language, just disable in e.g. some JSON-based configuration files (where credentials may occur, and so I don't want Copilot to scan those files).

I cannot find anything like this in configuration. Any chances to get it in the future?

**Comment by davecheney** [comment]
Open, or create, a file with the extension you don’t want copilot to offer completions on. Click the copilot icon, choose _disable_

**Comment by kam193** [comment]
Thanks, Dave - I know this feature, but unfortunately, it allows just global or language (extension) based disabling. I'd like to avoid so broad disabling.

E.g.: I could have a `local-config.json` and `test-fixture.json` files in the same workspace. Thought, I'd like to use Copilot in the second, but not in the first.

**Reply by laalamrajesh** [reply]
@kam193, can you confirm whether this issue has been resolved or not?

**Reply by kam193** [reply]
I'm not using Copilot at the moment, so I cannot confirm or not. The whole time I was using it, the case wasn't resolved.

**Comment by BrennerSpear** [comment]
How do I disable it for my .env files?

**Reply by ghost** [reply]
open a .env file, then click the copilot icon. it will say both `disable globally` and `disable for xxx`, click the second option and the icon should change color.   to get it back, just click the icon once again

**Reply by midnightcodr** [reply]
I found enable/disable `.extension` implementation to be very buggy. Say I have both `.env` and `.gitattributes` under my project. If I disable copilot for `.env` (which will also disable for `.gitattributes` because copilot classifies these two, or any files starting with a dot as `properties` files) then enable for `.gitattributes` later on, it will eventually enable `.env` as well - definitely not something I would expect.

**Comment by OmerFI** [comment]
I think there will be a quick option that disables/enables the copilot for the current file or current workspace

**Comment by dustinaleksiuk** [comment]
I'd really like to be able to disable Copilot for a given path wildcard that I specify. It is totally broken to the point of being unusable for styled-component tag functions in a .ts file. We name those files with a `.styled.ts` suffix.

**Comment by cchalfan1** [comment]
It would be really nice if there were commands like eslint e.g. `// @copilot disable-file`

**Reply by ZuoFuhong** [reply]
+1

**Reply by asheroto** [reply]
+1

**Reply by mengyun8715** [reply]
+1

**Reply by scemama** [reply]
+1

**Reply by bonnth80** [reply]
+1

**Reply by shalanah** [reply]
+1

**Reply by ekpono-gv** [reply]
+1

**Reply by Hery387** [reply]
+1

**Reply by vanloc1808** [reply]
+1

**Comment by atipugin** [comment]
👍🏻 for the feature, sometimes I just want to play with code without any suggestions from Copilot, but don't want to actually disable it per language.

**Comment by kdawgwilk** [comment]
This is becoming more of a security concern for various companies.  I think it would be prudent to support extension settings to add file globs that should be ignored that we can check into our repos settings.json to prevent any information from being sent to github servers such as `.env` contents

**Reply by danielfree** [reply]
+1 for regex file globs, maybe just some global settings for the plugin at least

**Comment by tomij-secapp** [comment]
A way to exclude certain files would be very needed. Preferably with an option to do that in a centralized manner.

**Reply by kevinferrero** [reply]
a ".copilotignore" file listing filenames and patterns to disable suggestions in would be awesome.

**Reply by atipugin** [reply]
+1 for .copilotignore. It could be utilized by other editors with copilot plugins (e.g. vim)

**Reply by JPustkuchen** [reply]
I think this is a good idea! Shouldn't this be an issue then (feature request) to be seen by the developers?

**Reply by kevinferrero** [reply]
Happy to submit a feature requests for this, but I haven't found the right spot.  The VS Code extension links to this community discussion area for Issues.  A search of copilotignore shows several repeats of the question.
https://github.com/orgs/community/discussions?discussions_q=copilotignore

The VS Code extension itself seems to name .copilotignore as a file format... but it doesn't appear to function, and I haven't found mention of it in any documentation.  It shows up in my list of file formats though.

![CleanShot 2023-04-20 at 13 31 53](https://user-images.githubusercontent.com/304900/233443622-2cb6eec2-a00e-4f71-a392-74124643e90a.png)

**Reply by fili** [reply]
fyi: https://github.com/orgs/community/discussions/10305

**Reply by okineadev** [reply]
> a ".copilotignore" file listing filenames and patterns to disable suggestions in would be awesome.

I suggest using the universal `.aiexclude` format, which is not tied to a specific company, application or AI, which is just suitable for everything - https://github.com/orgs/community/discussions/11254#discussioncomment-10461122

**Comment by timeasch** [comment]
Dňa po 27. 3. 2023 o 12:02 tomij-secapp ***@***.***>
napísala(a):

> A way to exclude certain files would be very needed. Preferably with an
> option to do that in a centralized manner.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/community/discussions/11254#discussioncomment-5438982>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/A6XYXKWIQLQLGTR4DCY5GG3W6FQULANCNFSM5N55X47A>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***
> com>
>
-- 
Schröderová

**Comment by LiamKarlMitchell** [comment]
This is kind of necessary feature per project to look at an ignore file.
.copilotignore
```
.env
settings.yaml
firebase_options.dart
```
You might want to ignore it in some files but use it in same file type elsewhere in same project e.g. populating test data in json files but for whatever reason using json file that might also have secrets you don't want to upload to internet.

**Reply by EricDuminil** [reply]
`.env` should be disabled by default. Opening VSCode + Copilot in any folder with an .env file is a security risk right now. Even if it just takes 3s to change the setting and prevent Copilot from checking .env, those 3s can be enough for credentials to be leaked.

**Comment by gmeligio** [comment]
I would like to have the option to disable it in "Untitled files" like it was reported here https://github.com/orgs/community/discussions/67236

Sometimes I need to check sensitive snippets, and I paste them in an editor. I can't use IDEs with Copilot because it can't be disabled. It's an overhead for me to be careful with what type of content I have on the clipboard and choose the text editor based on that.

**Comment by starball5** [comment]
Related on Stack Overflow: [How to Exclude Specific Files (like .env) from GitHub Copilot in VS Code?](https://stackoverflow.com/q/77780462/11107541)

**Comment by Superskyyy** [comment]
cant it just inherent from something like .gitignore

**Comment by asheroto** [comment]
Take a look at this. The dev just came out with it recently and it works great! Simply install the extension and create `.copilotignore` with glob patterns. 😊

https://github.com/orgs/community/discussions/10305#discussioncomment-8226775

https://github.com/mattickx/copilotignore-vscode

![image](https://github.com/community/community/assets/49938263/6c95aa71-1339-4094-9687-2f88e7da1059)

**Reply by Superskyyy** [reply]
great!

**Reply by spaiq** [reply]
Does this take care of security concerns i.e. stops copilot from reading these files and works for copilot chat as well?

**Comment by dmm2321** [comment]
If you all are using VSCode, you may want to look into the Profile feature. Your VSCode account and GitHub authorizations don't change, but it's as if you're using a different account so you can modify your settings and extensions to for specialized projects/uses. 

For example, from the original post, it looks like she may have a job handling security sensitive files. So she could create a "Work" profile to only include the necessary extension; she can choose to not install Copilot or any other extensions that may cause a problem/violation in her work. 

You can access your profiles from the File menu > Preferences. It switches over pretty fast and it's definitely faster than if you are repeatedly changing the same settings and extension access.

Here's the link to the VSCode documentation. Hope this helps you all out :)
https://code.visualstudio.com/docs/editor/profiles

**Reply by strudelPi** [reply]
I think the problem that folks are trying to solve is having the extension for work where files with possibly same extensions or extensions not recognized as a "language" in vscode might have sensitive data. I do not think this is a solution.

**Comment by dtinth** [comment]
Seems like this is now available in [**Copilot Business**](https://docs.github.com/en/enterprise-cloud@latest/copilot/copilot-business/about-github-copilot-business#understanding-the-differences-between-copilot-business-copilot-individual-and-copilot-enterprise)

![image](https://github.com/community/community/assets/193136/0c28c8c3-29e3-4194-af19-61d44bd4dd83)

https://docs.github.com/en/copilot/managing-github-copilot-in-your-organization/configuring-content-exclusions-for-github-copilot

**Reply by nabe1653** [reply]
Is this working on local files too?

I found the doc that saying set rule on repo too, but VSCode will work local file too.

We thought client will have settings, not repo(cloud) base rule.

**Reply by okkdev** [reply]
Super weird solution. What if my repo isn't on github?

**Reply by rupert-jung-mw** [reply]
Don't even have this setting available.

**Reply by jakemayfield** [reply]
I'm convinced that GitHub has no intention of making this feature available to individual, paying subscribers. I think I'll cancel and consider alternatives in its place until GitHub comes around. This stinks, plain and simple.

**Comment by okineadev** [comment]
Good afternoon, I suggest using the name `.aiexclude` instead of `.copilotignore` so that this file is suitable for everyone, both **Gemini** in Android Studio and **GitHub Copilot**

Link to `.aiexclude` file documentation - https://developer.android.com/studio/preview/gemini/aiexclude

**Reply by asheroto** [reply]
Thanks for that info. I can see that it's viable for Android Studio, but I'm having trouble finding the documentation that explains how GitHub Copilot honors it as well. Do you have that link?

**Reply by okineadev** [reply]
> Thanks for that info. I can see that it's viable for Android Studio, but I'm having trouble finding the documentation that explains how GitHub Copilot honors it as well. Do you have that link?

There is no such link, I came up with the initiative myself, I hope GitHub will see this and notice it, because it is really necessary to control which resources the AI can see and which cannot, because even accidentally confidential files can get into its training

**Reply by asheroto** [reply]
Ah I see what you're saying now. Indeed, that would be a good filename choice.

**Reply by wesleyboar** [reply]
**Excellent!** One suggestion at a time if we must. [I will help.](https://forum.cursor.com/t/questions-on-gitignore-cursorignore-cursorban/34713/6?u=wbtacc)

**Reply by janselv** [reply]
Might not be a good idea. One might want to restrict access to all/some files to Gemini but allow all files to Copilot

**Reply by okineadev** [reply]
> Might not be a good idea. One might want to restrict access to all/some files to Gemini but allow all files to Copilot

Give at least one case where this might be useful

And having a lot of configuration files for different AIs is bad, everything needs to be unified

**Comment by ruzzzz6312** [comment]
My API key leaked, long story short, full system is in a very controllable and closed environment no public repos, .env added to .gitignore, my only thought it leaked via Copilot as a suggestion to someone else in their IDE. 

Disabled it completely. Very very uncool that as an individual I can't disable certain files from copilot access.

**Reply by asheroto** [reply]
Right? This very thing scares me!

**Reply by ruzzzz6312** [reply]
disabled copilot

**Comment by oneconsciousness** [comment]
This fixed it for me
![image](https://github.com/user-attachments/assets/83c4beca-36c8-4d74-ab59-331f0bf8b72f)
Go to Settings => search "github.copilot.enable" => Add an entry {"key": "properties", "value": "false"}

**Reply by asheroto** [reply]
That's disabling for specific languages, but not for specific files.

`functions.php` <-- might be okay to read
`secrets.php` <-- not okay to read

**Reply by oneconsciousness** [reply]
Yeah, here's a *short-term workaround* for specific files. **Caveat**: these files lose language recognition (shouldn't be a huge problem for key/secrets? I may be biased). Thoughts? 
**Step 1:** Associate certain group of files I want to exclude to a custom language. Add `{"{.env,*.key,secrets.php}": "secret"}`
![image](https://github.com/user-attachments/assets/0b3aebfc-20bc-45dd-8de6-a1712414f333) 
**Step 2:** search `github.copilot.enable` => Add an entry `{"secret": "false"}`
![image](https://github.com/user-attachments/assets/3ef6bd66-64c9-49bf-ad93-f2a72fe49f02)
IMHO: Copilot should support this file exclusion more directly. Hope that happens soon!

**Reply by wesleyboar** [reply]
> [!WARNING]
> After doing this, Copilot can still **read** file and **its content**. I asked why it could, given my settings.
> > **GitHub Copilot**
> > Yes, I can still access the file `secrets.py` and provide assistance with it. The setting `"github.copilot.enable": { "secret": false }` only disables GitHub Copilot's code suggestions and completions for files associated with the "secret" language mode. It does not prevent me from accessing or viewing the file content.

<details><summary>Screenshots</summary>

| ai thinks it can access file | ai can read file content |
| - | - |
| <img width="960" alt="11254-11420433 workaround seems to fail 1" src="https://github.com/user-attachments/assets/90904f8a-a854-4162-8f89-3435e6a877f0" /> | <img width="960" alt="11254-11420433 workaround seems to fail 2" src="https://github.com/user-attachments/assets/56b9594e-3fdf-4cb8-9f8e-c9a0cc9a559a" /> |

<details><summary><em>Details</em></summary>

I closed and re-opened the workspace while removing and adding settings.

**Chat 1**
- me:
    > `#file:secrets.py` Can you read this file?
- ai:
    > Yes, I can read the file `/* secrets.py`. How can I assist you with it?

**Chat 1 (continued)**
- me:
    > Show me the first three lines from `#file:secrets.py`
- ai:
    > ```
    > ########################
    > # DJANGO
    > ########################
    > ```

**`settings.py`**
```
    "files.associations": {
        // ...
        "{secrets.py}": "secret"
    },
    "github.copilot.enable": {
        "secret": false
    },
```

**`secrets.py`**
```
########################
# DJANGO
########################

…
```

</details>

</details>

**Comment by QuadrupleA** [comment]
That Copilot gives individual users no way to avoid sending DB credentials, API keys, and other sensitive data to their all-seeing AI models, just by opening their IDE, is abusive and borderline criminal.

Let us exclude a file or two!

**Comment by wesleyboar** [comment]
[Cursor](https://www.cursor.com/) had a similar request. It seems to have just been achieved with [v0.46.x](https://www.cursor.com/changelog/agent-is-ready-and-ui-refresh) via [`.cursorignore` improvement & `.cursorindexingignore` feature](https://docs.cursor.com/context/ignore-files).

**Comment by abradaric** [comment]
wow, I can't believe this is locked behind business/enterprise plans. no one should leak secrets and api keys, free, pro, and everyone else!!!

**Comment by tiropictor** [comment]
Dammit. Was shocked to see my ENV data in Copilot replies recently. Guys, seriously, this is HUGE HOLE in security. API key, secrets - EVERYTHING is just leaking back and forth to goddamn FOUR EXTERNAL llm-models. Seriously? I've never added .env to git. I've put in to .gitignore, I've added it to .copilotingore, I've done ALL that is possible, but still - silicon brain watches me. 

User MUST have a possibility to RULE what ai is allowed to see and what is not. Otherwise it's just like walking the street without pants.

**Comment by tiropictor** [comment]


**Reply by sizhky** [reply]
simple and elegant!

**Comment by johannesjo** [comment]
Until this is fixed, I cannot use copilot at all.

**Comment by mkduer** [comment]
I don't know if others have found an answer for this, but thought I would share some experiments I tried.

Like others, I discovered that adding the .copilotignore file (same place as you might add a .gitignore and with similar content) was not sufficient.

If I prompted for an ignored file or its content, it would still be queried. Example, can you find this mock secret key in the ignored file, prompt would result with secret value and references to the file.

I restarted the IDE and repeated the experiment. This time it detected the .copilotignore and did not find the secret key or the file containing the secret.

**Comment by LiamKarlMitchell** [comment]
I ended up using Claude Code. Unsubscribe from this issue time.

**Comment by mkduer** [comment]
Yeah, agreed this thread is old. I take back my prior comment. It looks like .copilotignore is really an AI hallucination outside of the valid VS Code extension. 

Kind of blew my mind when I asked both Chat GPT and Claude Sonnet models if .copilotignore works and where to find official documentation. Both models made up references (e.g. Chat GPT provided this link https://docs.github.com/en/copilot/configurations/copilot-ignore which is not a real URL).

In Agent mode, they both insisted they would not reference files mentioned in the copilotignore, but sounds like following the offical documentation with the repository yaml setup is the safer approach along with some healthy skepticism.
